Userland/Applications/Mail/AccountHolder.h

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2021, Luke Wilde <lukew@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include "MailboxTreeModel.h"
+#include <AK/NonnullRefPtrVector.h>
+#include <AK/RefCounted.h>
+#include <AK/String.h>
+#include <LibIMAP/Objects.h>
+
+class BaseNode : public RefCounted<BaseNode> {
+public:
+    virtual ~BaseNode()
+    {
+    }
+};
+
+class MailboxNode;
+
+class AccountNode final : public BaseNode {
+public:
+    static NonnullRefPtr<AccountNode> create(String const& name)
+    {
+        return adopt_ref(*new AccountNode(name));
+    }
+
+    virtual ~AccountNode() override
+    {
+    }
+
+    void add_mailbox(NonnullRefPtr<MailboxNode> mailbox)
+    {
+        m_mailboxes.append(mailbox);
+    }
+
+    NonnullRefPtrVector<MailboxNode> const& mailboxes() const { return m_mailboxes; }
+
+    String const& name() const { return m_name; }
+
+private:
+    explicit AccountNode(String const& name)
+        : m_name(name)
+    {
+    }
+
+    String m_name;
+    NonnullRefPtrVector<MailboxNode> m_mailboxes;
+};
+
+class MailboxNode final : public BaseNode {
+public:
+    static NonnullRefPtr<MailboxNode> create(AccountNode const& associated_account, IMAP::ListItem const& mailbox)
+    {
+        return adopt_ref(*new MailboxNode(associated_account, mailbox));
+    }
+
+    virtual ~MailboxNode() override
+    {
+    }
+
+    AccountNode const& associated_account() const { return m_associated_account; }
+
+    String const& name() const { return m_mailbox.name; }
+    IMAP::ListItem const& mailbox() const { return m_mailbox; }
+
+private:
+    MailboxNode(AccountNode const& associated_account, IMAP::ListItem const& mailbox)
+        : m_associated_account(associated_account)
+        , m_mailbox(mailbox)
+    {
+    }
+
+    AccountNode const& m_associated_account;
+    IMAP::ListItem m_mailbox;
+};
```

Comment:
This should be a proper tree with subfolders and such. It's not just an account which owns a flat list of mailboxes. https://datatracker.ietf.org/doc/html/rfc3501#section-6.3.8

Reply:
Done :+1: